### ISSUE TITLE
Text rendering issue in symbol layer when map is rotated/tilted

### ISSUE THREAD

**Issue by westnordost** [maplibre/maplibre-native#2788]
**Describe the bug**

When the map is tilted or rotated, text labels in a symbol layer are sometimes not rendered at all or rendered in a warped way. (As seen in the video) the warping seems to persist when rotating back to tilt=0, rotation=0

https://github.com/user-attachments/assets/7c24179a-d3a1-4796-ab85-1cf8423dbea1

(This video was made with Android 9, Moto X4.)

https://github.com/user-attachments/assets/43990869-703a-44c4-837c-df16c7039cf8

![](https://github.com/user-attachments/assets/ea53e9be-28c9-441f-baa2-7b523ab082d0)

**To Reproduce**
Reproducible with https://github.com/streetcomplete/StreetComplete/releases/tag/v59.0-alpha2

Style affected:
```kotlin
SymbolLayer("pin-cluster-layer", SOURCE)
  .withFilter(all(
    gte(zoom(), 13f),
    lte(zoom(), CLUSTER_MAX_ZOOM),
    gt(toNumber(get("point_count")), 1)
  ))
  .withProperties(
    iconImage("cluster-circle"),
    iconSize(sum(literal(0.5f), division(log2(get("point_count")), literal(10f)))),
    textField(get("point_count")),
    textFont(arrayOf("Roboto Regular")),
    textOffset(arrayOf(0f, 0.1f)),
    textSize(sum(literal(15f), division(log2(get("point_count")), literal(1.5f)))),
    iconAllowOverlap(true),
    textAllowOverlap(true),
)
```

**Expected behavior**
Text renders correctly

**Platform information:**
 - Android, any version
 - Version 11.2.0

**Additional context**

Originally reported in https://github.com/streetcomplete/StreetComplete/issues/5859

So far, the same issue could **NOT** be reproduced for the following layer, even though it is also a symbol layer with text:

```kotlin
SymbolLayer("overlay-symbols", SOURCE)
  .withFilter(all(
    gte(zoom(), 17f),
    isPoint()
  ))
  .withProperties(
    iconImage(get("icon")),
    iconSize(interpolate(linear(), zoom(), stop(17, 0.5f), stop(19, 1f))),
    textField(get("label")),
    textFont(arrayOf("Roboto Regular")),
    textAnchor(Property.TEXT_ANCHOR_TOP),
    textOffset(switchCase(has("icon"), literal(arrayOf(0f, 1f)), literal(arrayOf(0f, 0f)))),
    textSize(16 * context.resources.configuration.fontScale),
    textColor(if (isNightMode) "#ccf" else "#124"),
    textHaloColor(if (isNightMode) "#2e2e48" else "#fff"),
    textHaloWidth(2.5f),
    iconColor(if (isNightMode) "#ccf" else "#124"),
    iconHaloColor(if (isNightMode) "#2e2e48" else "#fff"),
    iconHaloWidth(2.5f),
    textOptional(true),
    iconAllowOverlap(true),
    textAllowOverlap(step(zoom(), literal(false), stop(21, true))),
    symbolZOrder(Property.SYMBOL_Z_ORDER_SOURCE),
  )
```

This suggests that certain properties may trigger the issue and it can not be reproduced with just any text on any symbol layer.

**Comment by mtpdog** [comment]
I have the same issue. It started from the version 11.1.0. The latest stable version without this issue is 11.0.2-pre0.
According to my short tests 11.2.0 also have some additional weird tile loading issue (some tiles are blinking even when they are already loaded). Probably this is a related rendering bug.

**Comment by louwers** [comment]
cc @sjg-wdw

**Comment by louwers** [comment]
I tried to reproduce this in our test app. No luck so far...

https://github.com/maplibre/maplibre-native/commit/3cf9a8ea9aceb79b8b6edca09f1d0556378e4297

![image](https://github.com/user-attachments/assets/f9b4f66a-f93a-4b7c-9487-6358e926f187)

**Comment by mtpdog** [comment]
> Can you try removing the expressions which use logarithms?

According to my test this doesn't seem to be related to exponential interpolation. I have the same issue with linear interpolation and without interpolation at all.

**Comment by westnordost** [comment]
I removed everything from the layer style except this
```kotlin
        SymbolLayer("pin-cluster-layer", SOURCE)
            .withFilter(all(
                gte(zoom(), 13f),
                lte(zoom(), CLUSTER_MAX_ZOOM),
                gt(toNumber(get("point_count")), 1)
            ))
            .withProperties(
                textField(get("point_count")),
                iconAllowOverlap(true),
                textAllowOverlap(true),
            ),
```
and it is still reproducible with this.

Note that I wasn't able to reproduce this issue with things such as shop names or housenumbers yet

**Comment by louwers** [comment]
@westnordost Could you try? https://github.com/maplibre/maplibre-native/commit/3cf9a8ea9aceb79b8b6edca09f1d0556378e4297

**Comment by westnordost** [comment]
Hm, I didn't check out and setup this project on my computer yet. You want to see if it is maybe device-specific? After all, you already found that you cannot reproduce it with https://github.com/maplibre/maplibre-native/commit/3cf9a8ea9aceb79b8b6edca09f1d0556378e4297 . To see if it is device-specific, I think it would be quicker if you installed https://github.com/streetcomplete/StreetComplete/releases/tag/v59.0-alpha4 on your test device, zoomed out after the quests were downloaded until the clusters are visible and then rotate, push around and tilt the map to see if you have the same effect on your test device.

**Comment by westnordost** [comment]
@mtpdog are you using clustering?

**Comment by louwers** [comment]
Sure I can do that!

**Comment by westnordost** [comment]
Hm, weird. Note, though that it does occur not always on rotation, you have to push it around, zoom in and out, maybe tilt a bit. I didn't notice any specific thing that triggered it. I for example wasn't able to reproduce it on my Samsung S10e  with LineageOS (Android 13) in some village that had few quests (=small and few cluster bubbles) but this might as well have been coincidence.

Also, I was never able to reproduce on my phone what you can see in the first video posted in the issue. For me, the numbers only sometimes vanish, but never warp.

**Comment by westnordost** [comment]
Devices on which it was confirmed that it can be reproduced so far (I will update this post as new information from users comes in):

Warping text:
- Android 9, Moto X4

Missing text:
- Android 13, Samsung Galaxy S10e
- Android 14, unknown Motorola phone

**Comment by mnalis** [comment]
I also occasionally had warping on Android 14 (Samsung Galaxy S23+) like this:

https://github.com/user-attachments/assets/039d07ea-0bf5-4cef-871d-f14856e5fde3

**Comment by mtpdog** [comment]
> are you using clustering?

No, I don't. But I use custom fonts from my "glyphs" folder.

**Comment by mtpdog** [comment]
My suggestion is to examine changes made between versions 11.0.2-pre0 (the last working one) and 11.1.0 (this already includes warping).

**Comment by sjg-wdw** [comment]
A lot going on in this thread, but I think at the core is a problem with clustering.
Do we have any good clustering tests?

**Comment by mtpdog** [comment]
Just a short notice. According to my tests:
- this issue is **fixed** in **11.3.0-vulkan-pre0** 
- 11.4.0 still has this issue

**Comment by westnordost** [comment]
Testing with 11.3.0-vulkan-pre0, too, so far I can confirm that this issue does not occur with that version.

**Comment by pstanoev** [comment]
Happens on Samsung S23+ frequently. Happens when the map is tilted, rotated.

**Comment by mtpdog** [comment]
> Happens on Samsung S23+ frequently. Happens when the map is rotated.
> 
> Doesn't happen when the layer doesn't contain text-field. Starts happening after text field is added. The color of the stripes is matching the icon color (tried with multiple different icons, colors always matched the icon color).
> 
> It seems to be happening more when `iconAllowOverlap(true)` is added, not sure if it is correlated.
> 
> Update: After some experiments, I think it is happening when one of these is set to true: `text-allow-overlap = true` `icon-allow-overlap = true` `icon-ignore-placement = true`


Is it still not fixed in the newest versions? Have you checked?
I'm still on 11.0.2-pre because of this issue.

**Comment by sargunv** [comment]
Can confirm still happening on v11.5.1. This is on a Galaxy Z Fold 6 with the [OpenFreeMap Positron](https://tiles.openfreemap.org/styles/positron) style, though I've seen it with other styles too. 

https://github.com/user-attachments/assets/dedcd80f-9714-4be8-9f7c-bd15fc88cf7a

Repro here: https://github.com/sargunv/train-tracker-app/tree/2de043c7e2bd16f25e3a29374e0dfbafef0182b7

The same codebase/style with MapLibre iOS v6.7.1 doesn't have the issue.

**Comment by josxha** [comment]
This issue appears to be fixed in version 11.6.1.
Maybe it was similar to https://github.com/maplibre/maplibre-native/issues/2879?

**Comment by louwers** [comment]
Oh really, interesting! Please leave a comment if someone sees it again.

cc @TimSylvester @alasram
